core/src/worker/workflow/mod.rs

```diff
@@ -1003,6 +1021,38 @@ enum WFTReportStatus {
     /// [ActivationCompleteOutcome::WFTFailedDontReport]
     DropWft,
 }
+#[derive(Debug, Default)]
+struct BufferedTasks {
+    /// There should only be one buffered actual WFT at a time, since any new one will immediately
+    /// supersede any old one.
+    wft: Option<PermittedWFT>,
+    /// For query only tasks, multiple may be received concurrently and it's OK to buffer more
+    /// than one - however they should be dropped if, by the time we try to process them, we
+    /// have already processed a newer real WFT than the one the query was targeting (otherwise
+    /// we'd return data from the "future").
+    query_only_tasks: VecDeque<PermittedWFT>,
+}
+
+impl BufferedTasks {
+    fn buffer(&mut self, task: PermittedWFT) {
+        if task.work.is_query_only() {
+            self.query_only_tasks.push_back(task);
+        } else {
+            let _ = self.wft.insert(task);
+        }
+    }
+
+    fn has_tasks(&self) -> bool {
+        self.wft.is_some() || !self.query_only_tasks.is_empty()
+    }
+
+    /// Remove and return the next WFT from the buffer that should be applied
+    fn get_next_wft(&mut self) -> Option<PermittedWFT> {
+        self.wft
+            .take()
+            .or_else(|| self.query_only_tasks.pop_front())
+    }
```

Comment:
A comment about the ordering preference for WFT ahead of Query-only tasks would be good.

Reply:
Will add, and this reminds me I meant to test that we won't reply to queries that are "old" if we processed a newer WFT than they targeted.